src/hardware/driver/relay.rs

```diff
@@ -0,0 +1,242 @@
+    }
+
+    // set K1 to upper position
+    fn disengage_k1(&mut self) {
+        let mut mcp = self.gpio.try_lock().unwrap();
+        // set en high and en _n low in order to engage K1
+        mcp.set_gpio(self.k1_en.into(), Level::Low).unwrap();
+        mcp.set_gpio(self.k1_en_n.into(), Level::High).unwrap();
+    }
+
+    // set K1 to lower position
+    fn engage_k1(&mut self) {
+        let mut mcp = self.gpio.try_lock().unwrap();
+        mcp.set_gpio(self.k1_en.into(), Level::High).unwrap();
+        mcp.set_gpio(self.k1_en_n.into(), Level::Low).unwrap();
+    }
+}
+
+impl<I2C, E> sm::StateMachine<Relay<I2C>>
+where
+    I2C: WriteRead<Error = E> + Write<Error = E>,
+    E: Debug,
+{
+    /// Start relay enabling sequence. Returns the relay delay we need to wait for.
+    pub fn enable(&mut self) -> fugit::MillisDuration<u64> {
+        self.process_event(sm::Events::Enable).unwrap();
+        Relay::<I2C>::K0_DELAY // engage K0 first
+    }
+
+    pub fn disable(&mut self) -> fugit::MillisDuration<u64> {
+        self.process_event(sm::Events::Disable).unwrap();
+        Relay::<I2C>::K1_DELAY // engage K1 first
+    }
+
+    /// Handle a completed relay transition. Returns `Some(relay delay)` if we need to wait,
+    /// otherwise returns `None`.
+    pub fn handle_relay(&mut self) -> Option<fugit::MillisDuration<u64>> {
+        self.process_event(sm::Events::RelayDone).unwrap();
+        match *self.state() {
+            sm::States::EnableWaitK1 => Some(Relay::<I2C>::K1_DELAY), // disengage K1 second
+            sm::States::DisableWaitK0 => Some(Relay::<I2C>::K0_DELAY), // disengage K0 second
+            _ => None, // done, no delay needed
+        }
+    }
+}
+
+/// A SharedMcp can provide ownership of one of two sets of relays on Driver.
+/// Each set consists of two relays that control the state of a Driver output channel.
+/// The relays are controlled by toggeling pins on an MCP23008 I2C IO expander on the Driver board.
+/// Both sets use the same MCP23008 chip, hence the SharedMcp.
+pub struct SharedMcp<I2C1> {
+    mutex: spin::Mutex<Mcp230xx<I2C1, Mcp23008>>,
+}
+
+impl<I2C, E> SharedMcp<I2C>
+where
+    I2C: WriteRead<Error = E> + Write<Error = E>,
+    E: Debug,
+{
+    /// Construct a new shared MCP23008.
+    ///
+    /// # Args
+    /// * `mcp` - The MCP23008 peripheral to share.
+    pub fn new(mcp: Mcp230xx<I2C, Mcp23008>) -> Self {
+        Self {
+            mutex: spin::Mutex::new(mcp),
+        }
+    }
+
+    /// Allocate a set of relay pins on the MCP23008 and get the corresponding [Relay]
+    ///
+    /// # Args
+    /// * `ccdr`    - core clocks to construct a `delay` for flipflop control
+    /// * `ch`      - The Driver channel the relay pins are used for.
+    ///
+    /// # Returns
+    /// An instantiated [Relay] whose ownership can be transferred to other drivers.
+    pub fn obtain_relay(&'static self, ch: Channel) -> Relay<I2C> {
+        Relay::new(&self.mutex, ch)
+    }
```

Comment:
There is no protection against multiple instantiation of the same relay, right? They would have independent state machines and would not match reality.

Ryan was tracking ADC channel allocation status in https://github.com/quartiq/stabilizer/blob/d6df836b8be45496d7c2dca6289ce3b05f1c5346/src/hardware/shared_adc.rs and making sure they aren't allocated multiple times. But with ADC channels there is actually no problem allocating them twice (they don't have internal state). Hence it was removed there. Your case explicitly has a lot of internal state (the machine).